[CDAP-18627] Add put,get,getLatest,delete LevelDBTableCore APIs

wyzhang opened the pull request:
Why:
    Introduce put,get,delete APIs that operate on values at max version,
    so caller doesn't need to specify a version when it doesn't care.
    Introduce getLatest which returns the value with highest version.
    This can be used when caller doesn't know the version number of
    the latest value, but want to retrive the latest value.

    Introducing these APIs allows simplifying the logic on the caller side.
    For instance, a follow-up PR will make preview store to use these APIs.

chtyim commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableCore.java:
```diff
@@ -173,10 +173,73 @@ public void persist(Map<byte[], ? extends Map<byte[], byte[]>> changes, long ver
+  }
+
+  /**
+   * Read the latest (i.e. highest version) value at the target row and column.
```
When `tx` is provided, I believe it is reading the latest committed value, right? Better be specific in the doc.

wyzhang replied:
Yes. Updated the comment.

chtyim commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableCore.java:
```diff
@@ -173,10 +173,73 @@ public void persist(Map<byte[], ? extends Map<byte[], byte[]>> changes, long ver
+    try (DBIterator iterator = getDB().iterator()) {
+      iterator.seek(startKey);
+      while (iterator.hasNext()) {
+        Map.Entry<byte[], byte[]> entry = iterator.peekNext();
```
Why need to peek? Isn't that we can always just consume it by calling `iterator.next()`?

wyzhang replied:
We don't want to consume if it is the end key (which is an exclusive upper bound for the scan) or past the end key.

wyzhang replied:
In this case, either peek or consume is functionally correct, but using peek is more logically sound for the reason of not consuming past end key.

chtyim replied:
Why we don’t want to consume it? The iterator is local and just calling next makes the code simpler

wyzhang replied:
Done

chtyim commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableCore.java:
```diff
@@ -173,10 +173,73 @@ public void persist(Map<byte[], ? extends Map<byte[], byte[]>> changes, long ver
+          break;
+        }
+
+        // if this is part of a multi-row scan and we reach the next row, stop without advancing iterator
```
Is this comment accurate? This method is only getting one value, not a multi-row scan?

wyzhang replied:
Removed the comment.

chtyim commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableCore.java:
```diff
@@ -173,10 +173,73 @@ public void persist(Map<byte[], ? extends Map<byte[], byte[]>> changes, long ver
+        iterator.next();
+
+        // Determine if this KV is visible
+        if (tx != null && !tx.isVisible(kv.getTimestamp())) {
```
If tx is provided, it seems like this method will be returning the lowest instead of the highest committed (visible) version?

wyzhang replied:
Returning highest committed version, as values are sorted in version descending order to optimize for latest-version read (i.e. implemented in io.cdap.cdap.data2.dataset2.lib.table.leveldb.KeyValue.KeyComparator#compareTimestamps)

tivv commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableCore.java:
```diff
@@ -173,10 +173,73 @@ public void persist(Map<byte[], ? extends Map<byte[], byte[]>> changes, long ver
+   * Write the value at the target row and column with the max version {@link KeyValue.LATEST_TIMESTAMP}.
+   * as a result it hides any value written with equal or smaller version.
+   */
+  public void put(byte[] row, byte[] column, byte[] value) throws IOException {
```
Why is it needed? What would happen if there are multiple put calls? This value is also used in querying, is it safe to use it for writing?

wyzhang replied:
Used by callers who don't care about versioning values. They can put(row, col, val). Each put call overwrites values by previous put calls.

tivv replied:
Can we use some other magic value? This LATEST_TIMESTAMP is used to define ranges and I am worrying that we may break some inclusive/exclusive queries that use LATEST_TIMESTAMP on it's bounds and don't care much about edge inclusion. e.g. can we use `0` ?

wyzhang replied:
LATEST_TIMESTAMP is the right one to use, existing code is built on the assumption latest_timestamp corresponding to the most recent version (i.e. the version of value to be returned upon read). This ensure that any write overwrites existing value.

tivv commented on cdap-data-fabric/src/main/java/io/cdap/cdap/data2/dataset2/lib/table/leveldb/LevelDBTableCore.java:
```diff
@@ -330,9 +393,33 @@ public void close() {
+  /**
+   * Delete the cell at specified row and column with max version {@link KeyValue.LATEST_TIMESTAMP}.
+   */
+  public void delete(byte[] row, byte[] column) throws IOException {
```
This call is pretty confusing. If I'd to see the call, I'd expect it to delete all versions, not just some special version. What is the use case?

wyzhang replied:
Used by caller who don't care about reading/writing/deleting value to a particular version.  

For instance, preview manager would 
put (row, col, val) 
then
get(row, col) and delete(row, col) 

All of the above operate on a default version, which is max version. Otherwise the code is a bit tedious like
put(row, col, val, 1) 
then 
get(row, col, 1) and delete(row, col, 1) 
Specifying version is unnecessary and may also expose possibility of making errors when a wrong version gets specified. Wish java allows setting default value for parameter like C++, that might help with clarify the confusion you have when see the function signature.

To delete all versions, method would better to be named deleteAllVersions() and we should have corresponding getAllVersions(). At the moment, there is no strong need for deleteAllVersions() and getAllVersions()

tivv replied:
Still please name it as `deleteDefaultVersion`. Usually when there are versions a delete call without specifying version removes all of it, not some "magic" version

wyzhang replied:
Done. Updated all APIs to [get|put|delete]DefaultVersion

wyzhang commented:
For your reference, PR for preview store to use new APIs introduced here, so you can see how these APIs are going to used.

wyzhang commented:
https://github.com/cdapio/cdap/pull/13768

wyzhang commented:
PTAL. Added a test case for scan to ensure writing at default (i.e. max version) doesn't cause problem for scan.

wyzhang committed:
[CDAP-18627] Add put,get,getLatest,delete LevelDBTableCore APIs

Why:
Introduce put,get,delete APIs that operate on values at max version,
so caller doesn't need to specify a version when it doesn't care.
Introduce getLatest which returns the value with highest version.
This can be used when caller doesn't know the version number of
the latest value, but want to retrive the latest value.

Introducing these APIs allows simplifying the logic on the caller side.
For instance, a follow-up PR will make preview store to use these APIs.
